docs/source/miscellaneous/doc/adr/0016-use-tickets-encoding-for-transactions.md

```diff
@@ -499,7 +499,193 @@ the worker pool has a finite size and even with the full batching algorithm in t
+|-----:|--------:|--------:|-----------------------------------------------------------------------:|
+| 0x6d |    0x00 |      -1 | 0x7b2273657175656e6365223a31333030372c22626f756e64223a353031333030377d |
+| 0x6d |  0xa711 |      -1 |                                     0x7b2276657...3334227d (216 bytes) |
+| 0x6d |  0xb2cf |      -1 |                                     0x7b2276657...3930227d (380 bytes) |
+
+The sequence IDs are encoded as a ``STRING``. The column IDs are ``VAR_LONG`` encoded.
+
+A consequence of this in Cassandra KVS is that all entries for a single coordination sequence will be stored in the
+same replication group. However, this is expected to be acceptable as we don't range scan this table, and the number
+of values to be written is small.
+
+#### Determining Transaction Schema Versions
+
+We still need to retain the ability to read from the old transactions table, when checking if/when values written
+before the transactions2 migration occurred. Thus, we replace the old ``TransactionService`` (that only reads/writes
+from transactions1) with one that delegates between multiple TransactionServices (concretely, just V1 and V2 at this
+time) depending on the specific timestamps involved in calls.
+
+The mechanism for determining which ``TransactionService`` to use goes through the aforementioned
+``CoordinationService``. Nodes agree on a ``TimestampPartitioningMap``, a mapping of timestamps to transaction schema
+versions, using the coordination service; this map is always the current value of the coordination store.
+This is a range map, as we want to support rollbacks easily. The map also is guaranteed to span the
+ranges ``[1, +∞)`` and be connected; note that this does not mean future behaviour is fixed, since in practice this map
+is read together with a validity bound. Thus, even though the map will contain as a key some range ``[C, +∞)`` for a
+constant ``C``, it should be valid up to some point ``V > C`` - and behaviour at timestamps after ``V`` may subsequently
+be changed.
+
+Note that the ranges are based on the start timestamp; since we accessed the original ``TransactionService`` via the
+start timestamp, we will use the start timestamp against the range map to find out the behaviour at that point.
+To give a concrete example, if our range map is ``{[1, 5000) = 1, [5000, +∞) = 2}`` and we want to commit a transaction
+that started at timestamp 4990 and finished at 5010, it is still stored under schema version 1 (even if other
+transactions that started later may have already written to schema version 2 - if they started at 5000 and finished at
+5001, for instance).
+
+We attempt to read the latest version of the range map, and if our timestamp falls within the validity bound, we
+retrieve the version. If it does not, we submit an identity transformation, which extends the validity bound of the
+current value. We then read the value again, and keep trying until we know what version to use. If we read a version
```

Comment:
Added a line about this:

>  Our
transform may be rejected if another transform has taken place and changed the value from underneath us; however, because all transforms push the bound forward, we will eventually have a value that is valid for our timestamp.